Add GeneratorUtils module, add exercise: circular-buffer, add exercise: darts

vyu opened the pull request:
Sorry for adding three things at once, but I'm putting these commits in a single PR because there are inter-dependencies that could be more confusing if I submitted them separately. :frowning: But I suggest looking at each commit individually.

The first commit adds the `GeneratorUtils` module (and associated scripts) which is used by generators in the next two commits to create `runtests.jl` files for the circular-buffer and darts exercises. The last commit is the largest and not very important—it's just a directory of `runtests.jl` files generated by the `generate-defaults.jl` script to demonstrate the default settings of `GeneratorUtils`, and these files should not be used directly for an exercise's `runtests.jl`.

Included is the bash script `bin/diff-generated-runtests.sh`, which checks each generator that uses `GeneratorUtils` (currently only circular-buffer and darts) by downloading the latest `canonical-data.json` from https://github.com/exercism/problem-specifications/ and `diff`ing the generated `runtests.jl` with the existing one, returning the `diff` with exit status 1 if they are not identical.

For the circular-buffer exercise, I've added bonus tasks (in `.meta/hints.md`) that substantially expands its scope and I've assigned the difficulty of 8 based on this. Students can enable the associated bonus tests by adding the line `enable_bonus_tests = true`.

For the darts exercise, the generator file (`generators/darts.jl`) is just a symbolic link to the template file (`generators/GeneratorUtils/default_template.jl`) because it needs no changes to the template.

cmcaine commented:
configlet already requires problem-specifications to be in a dir adjacent
the root. Could use that rather than downloading on demand?

Maybe things would be clearer if you gave GeneratorUtils src and tests
subdirs and generally made it a real package?



On Fri, 26 Jun 2020, 07:38 Vincent Yu, <notifications@github.com> wrote:

> Sorry for adding three things at once, but I'm putting these commits in a
> single PR because there are inter-dependencies that could be more confusing
> if I submitted them separately. 😦 But I suggest looking at each commit
> individually.
>
> The first commit adds the GeneratorUtils module (and associated scripts)
> which is used by generators in the next two commits to create runtests.jl
> files for the circular-buffer and darts exercises. The last commit is the
> largest and not very important—it's just a directory of runtests.jl files
> generated by the generate-defaults.jl script to demonstrate the default
> settings of GeneratorUtils, and these files should not be used directly
> for an exercise's runtests.jl.
>
> Included is the bash script bin/diff-generated-runtests.sh, which checks
> each generator that uses GeneratorUtils (currently only circular-buffer
> and darts) by downloading the latest canonical-data.json from
> https://github.com/exercism/problem-specifications/ and diffing the
> generated runtests.jl with the existing one, returning the diff with exit
> status 1 if they are not identical.
>
> For the circular-buffer exercise, I've added bonus tasks (in
> .meta/hints.md) that substantially expands its scope and I've assigned
> the difficulty of 8 based on this. Students can enable the associated bonus
> tests by adding the line enable_bonus_tests = true.
>
> For the darts exercise, the generator file (generators/darts.jl) is just
> a symbolic link to the template file (
> generators/GeneratorUtils/default_template.jl) because it needs no
> changes to the template.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/exercism/julia/pull/251
> Commit Summary
>
>    - Add GeneratorUtils module and associated scripts
>    - Add exercise: circular-buffer
>    - Add exercise: darts
>    - Add generators/GeneratorUtils/generated_defaults/
>
> File Changes
>
>    - *A* bin/diff-generated-runtests.sh
>    <https://github.com/exercism/julia/pull/251/files#diff-2e714320ed6b6477117d85aff7cf11fb>
>    (34)
>    - *M* config.json
>    <https://github.com/exercism/julia/pull/251/files#diff-06b2d3b23dce96e1619d2b53d6c947ec>
>    (23)
>    - *A* exercises/circular-buffer/.meta/hints.md
>    <https://github.com/exercism/julia/pull/251/files#diff-a1c054afd98b4712dbaf76b066656a2b>
>    (70)
>    - *A* exercises/circular-buffer/README.md
>    <https://github.com/exercism/julia/pull/251/files#diff-4f8eaaa8136b1d250d78565d69e04e24>
>    (132)
>    - *A* exercises/circular-buffer/circular-buffer.ipynb
>    <https://github.com/exercism/julia/pull/251/files#diff-625d943735d8308c12861df021940d62>
>    (68)
>    - *A* exercises/circular-buffer/circular-buffer.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-47f3ff9143deb176547a7b54e29ada84>
>    (19)
>    - *A* exercises/circular-buffer/example.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-4e1456de0cd65eeacc543b4ac2aa0ba2>
>    (177)
>    - *A* exercises/circular-buffer/runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-6c58311d3f3ae077e1cae7af07745819>
>    (356)
>    - *A* exercises/darts/README.md
>    <https://github.com/exercism/julia/pull/251/files#diff-9f7a6b789b4767adfc55d656d8a06388>
>    (26)
>    - *A* exercises/darts/darts.ipynb
>    <https://github.com/exercism/julia/pull/251/files#diff-b34b99beace2b113c7f423dc237153f9>
>    (68)
>    - *A* exercises/darts/darts.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-3ba6289b21a41d49ed401d1915e8e3c4>
>    (3)
>    - *A* exercises/darts/example.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-b71c1d868e5cb6ab1065866c6130cdeb>
>    (9)
>    - *A* exercises/darts/runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-3701b768d8dd82e8328bdf67a88e559b>
>    (59)
>    - *A* generators/GeneratorUtils/GeneratorUtils.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-b2f1430ce559b81ddb6f5adca178b213>
>    (272)
>    - *A* generators/GeneratorUtils/default_template.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-6f477369c3f3c5b1eb8d24e5359999d6>
>    (38)
>    - *A* generators/GeneratorUtils/generate_defaults.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-1e1c3819edf5ab2912af2e8d02345faf>
>    (63)
>    - *A* generators/GeneratorUtils/generated_defaults/acronym_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-7c5d3c2a26f1d6c3fbd39b1ac2b447d1>
>    (37)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/affine-cipher_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-2d0e8c2dbe3140823218b9a2ebcc3b04>
>    (68)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/all-your-base_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-10e1ce0b0ed236bea17964e8c78b18fd>
>    (104)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/allergies_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-ee63510325f8ad9c89ce4bcc77cab9c2>
>    (187)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/alphametics_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-c9aa2da5b1269528f30a66df8ca15fde>
>    (40)
>    - *A* generators/GeneratorUtils/generated_defaults/anagram_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-0d9708af23167f50171efc716b34e3d9>
>    (63)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/armstrong-numbers_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-7c00adb84a86d6246ae88fd343c31ccb>
>    (43)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/atbash-cipher_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-faa4d155488c969343968864fa49a305>
>    (62)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/beer-song_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-1ad0a43e523540788b1bf11b8cc67ccb>
>    (41)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/binary-search-tree_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-5428c7fd6511acabb1bd83eea5e79958>
>    (75)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/binary-search_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-14f338e03427b2339760d2be5ed03a0a>
>    (65)
>    - *A* generators/GeneratorUtils/generated_defaults/binary_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-c2b8d8c24297aa5f078fd42f8199117b>
>    (67)
>    - *A* generators/GeneratorUtils/generated_defaults/bob_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-9add4b927ebab4355114521e9eaae896>
>    (107)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/book-store_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-b93fc3861ac94dd9fb397bdff6f98480>
>    (55)
>    - *A* generators/GeneratorUtils/generated_defaults/bowling_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-7953848ac96a06a7689fcde713102c21>
>    (148)
>    - *A* generators/GeneratorUtils/generated_defaults/change_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-985a5e9a2f85382bdd07eb185df692b6>
>    (58)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/circular-buffer_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-84dc28c45531dec254a42169bb322bcf>
>    (88)
>    - *A* generators/GeneratorUtils/generated_defaults/clock_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-6b41800260aa41851215e88666ea9fde>
>    (183)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/collatz-conjecture_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-b70b1ee451b82fe76789d0ae923f3295>
>    (31)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/complex-numbers_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-2b1b6752835e864ad1f5c5cc2f2eddd5>
>    (139)
>    - *A* generators/GeneratorUtils/generated_defaults/connect_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-2c4c4ee170a6ab6baa92339431dab5c0>
>    (47)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/crypto-square_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-f48df79c925c3a586e2460da1ee14a77>
>    (35)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/custom-set_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-c48b6f805e900d3ea8dbb375aedfaaf9>
>    (160)
>    - *A* generators/GeneratorUtils/generated_defaults/darts_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-e0ade5855480ea1639edcbc31a6c1a30>
>    (64)
>    - *A* generators/GeneratorUtils/generated_defaults/diamond_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-78cf8d3b0caf22800f54a740d7882b32>
>    (46)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/difference-of-squares_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-1d3d2606e2832ee602cd3b2aa2d57b45>
>    (43)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/diffie-hellman_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-18151a063b90c8d68a95aee189448693>
>    (43)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/dnd-character_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-62aae95bc4ec6f46881b48afb90e9789>
>    (89)
>    - *A* generators/GeneratorUtils/generated_defaults/dominoes_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-10bd1104a0abb556222aa00910b2444b>
>    (92)
>    - *A* generators/GeneratorUtils/generated_defaults/etl_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-b85cbd5cf0d25e35186b8f9925a72f15>
>    (22)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/flatten-array_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-de6e2a4997dede766e1dff5435b447a1>
>    (31)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/food-chain_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-074c36b85997e5c0e2c1df5d110e84bf>
>    (52)
>    - *A* generators/GeneratorUtils/generated_defaults/forth_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-ead62f1cf8ecb54aaeae769a4cef7e89>
>    (188)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/gigasecond_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-46c5dc3558854a7f52e28fffbf6e23bf>
>    (41)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/go-counting_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-6670405ba4a1cb44b54d51be5a37eefb>
>    (56)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/grade-school_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-ac176cd5740d785bc689a8adddf76df7>
>    (41)
>    - *A* generators/GeneratorUtils/generated_defaults/grains_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-25d2b6d4a51a521dd5ef7373efc22216>
>    (50)
>    - *A* generators/GeneratorUtils/generated_defaults/grep_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-c0a9718262fba46b71520be3c02241d3>
>    (141)
>    - *A* generators/GeneratorUtils/generated_defaults/hamming_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-adb41dbb9e2d8b49e0e924f9d6d7c370>
>    (54)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/hello-world_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-924fc9cf363f58eac0c0db0e4e816bd9>
>    (11)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/high-scores_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-4ec8c7eed8a343210f9e606ddaa884a8>
>    (47)
>    - *A* generators/GeneratorUtils/generated_defaults/house_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-9333f29084562f337f2bc206530c0228>
>    (64)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/isbn-verifier_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-4f20143d6dcfb0a207362178a867fea8>
>    (81)
>    - *A* generators/GeneratorUtils/generated_defaults/isogram_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-def0ddf6da7889872cb8f868f7ecaa10>
>    (54)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/kindergarten-garden_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-3434414f20f28060648e6abfc6d9c38f>
>    (42)
>    - *A* generators/GeneratorUtils/generated_defaults/knapsack_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-946f37d03e192cea850519274301d48a>
>    (40)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/largest-series-product_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-0ced24ad341f4a4674f6387419af11bd>
>    (80)
>    - *A* generators/GeneratorUtils/generated_defaults/leap_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-7ff34a38fc7c73c3d85d04663d0d60df>
>    (43)
>    - *A* generators/GeneratorUtils/generated_defaults/list-ops_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-2ed1434f07c72029312765d9b7d74019>
>    (101)
>    - *A* generators/GeneratorUtils/generated_defaults/luhn_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-379c1babe45f0e7ae2ab94d4e5dc7b1e>
>    (79)
>    - *A* generators/GeneratorUtils/generated_defaults/markdown_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-b566e0764516950536badc2a587686e3>
>    (64)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/matching-brackets_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-e5a00471d9a099188a1d5227f5129932>
>    (75)
>    - *A* generators/GeneratorUtils/generated_defaults/matrix_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-ee35063244916ca9bafe80c0dbce9ed8>
>    (39)
>    - *A* generators/GeneratorUtils/generated_defaults/meetup_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-cbf1e7626341e576db18046d7cef5163>
>    (387)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/micro-blog_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-74f4c2721616982ea471f48dffcea273>
>    (66)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/minesweeper_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-050cae1e3b9363c76c563248fb1c8556>
>    (63)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/nth-prime_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-519d3a6c0ae494e915dba8434685924b>
>    (27)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/nucleotide-count_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-6651854610f26622f0351b864864bc48>
>    (25)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/ocr-numbers_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-9383fb2181f2d35686347d97ec11c684>
>    (61)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/palindrome-products_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-b18c651323fbb5e43e3e806fc00911b0>
>    (55)
>    - *A* generators/GeneratorUtils/generated_defaults/pangram_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-033fa655b3bcf6623484bb4bebd0dfee>
>    (53)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/pascals-triangle_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-375bcaf262474b7fc50797b3c32059a5>
>    (40)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/perfect-numbers_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-7f4041ac98eff5c809d220a6c56ae290>
>    (58)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/phone-number_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-1a1aa774964faf928f0f014471b57754>
>    (64)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/pig-latin_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-8090aea5a9d9a88677201a7ff65f22f4>
>    (91)
>    - *A* generators/GeneratorUtils/generated_defaults/poker_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-b110b94e07d60c3148e454c9c23da968>
>    (119)
>    - *A* generators/GeneratorUtils/generated_defaults/pov_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-d44a461fc2c0c381e947b3de6dac907e>
>    (58)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/prime-factors_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-9e113e761f1927bf3de4e72271d17714>
>    (31)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/protein-translation_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-26b2185218c325f2e6619614c10fe8ab>
>    (79)
>    - *A* generators/GeneratorUtils/generated_defaults/proverb_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-1569da1029f88a526d588160454997b4>
>    (38)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/pythagorean-triplet_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-faa14beb16c4765bd94ab69419f29a10>
>    (35)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/queen-attack_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-f7fe215f1c4addf3fb30ff0eb14ef102>
>    (60)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/rail-fence-cipher_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-e63b5cb8603eeffcb5545aff15dd7dae>
>    (37)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/raindrops_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-e557c0440f19354cd3bc7afbb28064cf>
>    (79)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/rational-numbers_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-5ea1f229585e801be100c7d23aed1739>
>    (154)
>    - *A* generators/GeneratorUtils/generated_defaults/react_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-e0745728a4ed76d8a3fbbd1e7b44f283>
>    (107)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/rectangles_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-9d4684bdba3cbc0d819cff3137bbf01b>
>    (65)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/resistor-color-duo_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-dfd71a0ae215ac9b5d68e1987aba3d0d>
>    (27)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/resistor-color-trio_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-b9d1e6d6b573aed777b830493f64a45d>
>    (27)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/resistor-color_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-1648bd74eaa5cb70f17aeb813a9c5de1>
>    (23)
>    - *A* generators/GeneratorUtils/generated_defaults/rest-api_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-0e068915cfcc6a0debe5d998c7f7e21c>
>    (51)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/reverse-string_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-eb7202ed27ae7f8019510b8a8d8f02e8>
>    (36)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/rna-transcription_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-980ef7b3bf593b9a3388e0f4bc60f242>
>    (31)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/robot-simulator_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-34cc679dfb50ad8c550e7d29f5265182>
>    (85)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/roman-numerals_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-241044fdc39e2c8ecac41d624e5f3c7a>
>    (83)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/rotational-cipher_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-df05db7e04a55ca0c150ce7be95cc284>
>    (45)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/run-length-encoding_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-484050379ab9f5211fca4f06528fbaaf>
>    (55)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/saddle-points_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-1911d87b1f70c2d29f09a258ce490a3e>
>    (48)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/satellite_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-99f98b7ce5b63ff04c0d414ad713c0f2>
>    (31)
>    - *A* generators/GeneratorUtils/generated_defaults/say_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-e6d2ccdc7318bb9993865284e58339f4>
>    (73)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/scale-generator_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-97b783162da8081a5ee30ece8f31b8c2>
>    (64)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/scrabble-score_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-e0d4f5daaa678596ee49c0e8412be397>
>    (51)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/secret-handshake_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-865b16d87ba7738e0f83045053559357>
>    (58)
>    - *A* generators/GeneratorUtils/generated_defaults/series_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-5773c9c563e9334f60810d7dda2316a3>
>    (47)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/sgf-parsing_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-cde2490f53d51d0cde20884acaa6efa8>
>    (59)
>    - *A* generators/GeneratorUtils/generated_defaults/sieve_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-d45269a7e7663c8469d366550b995091>
>    (27)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/simple-cipher_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-902356f15183f94c7776946826645b87>
>    (58)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/space-age_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-91c726d339228d0455636978f1026af8>
>    (39)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/spiral-matrix_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-0915fc4976c7c1afb94402e21f8a6789>
>    (31)
>    - *A* generators/GeneratorUtils/generated_defaults/sublist_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-3b1b2da0080f49331562cd2503b60312>
>    (90)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/sum-of-multiples_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-f873e570ef545cb6325915d6ea963bc7>
>    (71)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/tournament_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-248581666a20597058c30b1e52eed827>
>    (61)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/transpose_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-f8efe281bddf2cb72140fdd4da6a490d>
>    (58)
>    - *A* generators/GeneratorUtils/generated_defaults/triangle_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-d981772670bb0afef8eb0376018bbdb2>
>    (91)
>    - *A* generators/GeneratorUtils/generated_defaults/trinary_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-53ebbef73619cb20b7d9ad7e3c74e8c1>
>    (43)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/twelve-days_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-4cbe3031633a6a1a8c64d5a669c2db65>
>    (65)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/two-bucket_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-e6c22a7c6d57c8d5db412cb3594491aa>
>    (31)
>    - *A* generators/GeneratorUtils/generated_defaults/two-fer_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-4e838eb947486b3c78e1708ef9cd3248>
>    (19)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/variable-length-quantity_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-178f6dbd81d28761f49f02d6eecb89d0>
>    (99)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/word-count_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-4fb9c2357adf733c053c86b141169650>
>    (65)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/word-search_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-08042afb8beef9e10a14b3fd91cdffee>
>    (92)
>    - *A* generators/GeneratorUtils/generated_defaults/wordy_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-3cc988552d4a2c130fa536c76698bb87>
>    (106)
>    - *A* generators/GeneratorUtils/generated_defaults/yacht_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-e0580600282bd8409ef1333b12b8f1f0>
>    (132)
>    - *A*
>    generators/GeneratorUtils/generated_defaults/zebra-puzzle_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-9f05cb966447ea66950dd7108f7e2d57>
>    (15)
>    - *A* generators/GeneratorUtils/generated_defaults/zipper_runtests.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-40501f65e4dca6a693f86b5d6aa39b38>
>    (76)
>    - *M* generators/Manifest.toml
>    <https://github.com/exercism/julia/pull/251/files#diff-dd789d987829e515f041c61c06f41170>
>    (10)
>    - *M* generators/Project.toml
>    <https://github.com/exercism/julia/pull/251/files#diff-468263e2fe954044a3128c61bd3e5694>
>    (1)
>    - *A* generators/circular-buffer.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-8ec35cf27411fcc40a33577c991f5ebd>
>    (311)
>    - *A* generators/darts.jl
>    <https://github.com/exercism/julia/pull/251/files#diff-df0b818f8734dce4673ed37b98b11158>
>    (1)
>
> Patch Links:
>
>    - https://github.com/exercism/julia/pull/251.patch
>    - https://github.com/exercism/julia/pull/251.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/julia/pull/251>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABNZA6ORHFBU73NY32UAGMLRYQ65BANCNFSM4OJBG6LQ>
> .)
>

vyu commented:
> configlet already requires problem-specifications to be in a dir adjacent the root. Could use that rather than downloading on demand?

Oh, I didn't realize that! Good idea, I've updated the script to use that.

> Maybe things would be clearer if you gave GeneratorUtils src and tests subdirs and generally made it a real package?

Yeah, I was unsatisfied with the way I organized it too. I took your suggestion and wrapped GeneratorUtils into a local package (albeit without tests), and moved some files into an `example/` dir there. It's now a part of the `generators/` project environment and no longer requires modifying `LOAD_PATH`. I think it looks much better now.

Two commits added.

cmcaine commented on generators/GeneratorUtils/GeneratorUtils.jl:
```diff
@@ -0,0 +1,272 @@
+# See default_template.jl for an example using this module.
```
I have not reviewed this.

cmcaine commented:
And thanks for submitting this PR! It mostly looks good :D

vyu commented on exercises/darts/example.jl:
```diff
@@ -0,0 +1,9 @@
+const score_table = (1 => 10, 5 => 5, 10 => 1)
+
+function score(x::Real, y::Real)
```
Would `(x, y)` be a better description than `[x, y]` in the docstring?

cmcaine replied:
Perhaps. Edited.

vyu replied:
Added to latest commit.

vyu commented:
Thanks for the review! Regarding circular-buffer, what do you think about removing all the bonus tasks and tests? If it's not a good extension to the exercise, I don't mind removing it.

vyu commented on generators/GeneratorUtils/default_template.jl:
```diff
@@ -0,0 +1,38 @@
+
+# Modify as desired.
+const IO_IN = stdin
+const IO_OUT = stdout
```
Nothing. I thought they looked better, but in my usual I/O workflow in notebooks I actually have global mutables `io_in = [...]` and `io_out = [...]` that I change to whatever is most convenient at the time. Would you recommend that, or do you mean you would go further to remove `io_in` and `io_out` and directly put in `stdin` and `stdout` where appropriate?

cmcaine replied:
If you find it useful then by all means leave it in. I can see the purpose for notebooks where you don't have a proper stdin or stdout, but I don't personally see why I would be editing those otherwise, but I don't see everything :)

vyu replied:
Okay, let's leave as-is for now.

vyu commented on exercises/circular-buffer/.meta/hints.md:
```diff
@@ -0,0 +1,70 @@
+[@which](https://docs.julialang.org/en/v1/stdlib/InteractiveUtils/#InteractiveUtils.@which)
+macro to identify the specific method to which a function call is dispatched. It also shows
+you the file and line number at which that method is defined (in a Jupyter Notebook through
+IJulia, it even gives you a link to the relevant code on GitHub).
```
I've never used `@edit` but it looks useful. Would you mind writing the mention?

cmcaine replied:
```suggestion
IJulia, it even gives you a link to the relevant code on GitHub).

If you are working at the REPL, you may prefer to use the [`@edit`](https://docs.julialang.org/en/v1/stdlib/InteractiveUtils/#InteractiveUtils.@edit) macro to open the relevant file and line in your default text editor.
```

vyu replied:
Thanks, added!

vyu commented on exercises/circular-buffer/example.jl:
```diff
@@ -0,0 +1,177 @@
+@inline function _dataindex(cb::CircularBuffer, i::Integer)
+    @boundscheck checkbounds(cb, i)
+    j = cb.lastindex - length(cb) + i
+    return j < 1 ? j + length(cb.data) : j
```
I ended up with this here and in similar code because `mod1` was too big a performance drain (you can tell I was trying for performance with the bounds annotations). The current overhead over array access is negligible, but benchmarking showed that `mod1` was causing something like a 10x slowdown, I guess because of a division.

cmcaine replied:
Fair enough :)

vyu replied:
I've added a comment in the code, though, because now that you've pointed it out I see it is confusing.

vyu commented on exercises/circular-buffer/example.jl:
```diff
@@ -0,0 +1,177 @@
+
+export CircularBuffer, capacity, isfull
+
+const OVERWRITE = false
```
Every few minutes, because I was switching between the generated tests and the tests from DataStructures.jl where overwriting was the default. :smile: But you're right, I should've taken this out afterward and I've made the edit.

vyu commented on exercises/circular-buffer/.meta/hints.md:
```diff
@@ -0,0 +1,70 @@
+- `Base.empty!(cb::CircularBuffer)`: Remove all elements from `cb`, then return the empty
+  `cb`.
+
+## Bonus tasks
```
Okay, I've added your suggestion and de-emphasized the bonus tasks.

vyu commented on exercises/darts/darts.jl:
```diff
@@ -0,0 +1,3 @@
+function score(x::Real, y::Real)
```
Okay.

vyu commented on exercises/circular-buffer/example.jl:
```diff
@@ -0,0 +1,177 @@
+    CircularBuffer(v::Vector{T}) -> CircularBuffer{T}
+
+Construct a `CircularBuffer{T}` using `v` as the backing data without copying. Its capacity
+and length are equal to `length(v)`.
```
Good point. I looked around Julia Base and it looks like the norm there at least is to organize things to avoid documenting both the type and an inner constructor at the same time. In any case, I've now massively simplified the constructors into a single inner constructor, since the rest are unnecessary for the exercise as you pointed out.

vyu commented on exercises/circular-buffer/example.jl:
```diff
@@ -0,0 +1,177 @@
+
+Construct a `CircularBuffer{T}` that stores at most `capacity` elements. If `undef` is
+provided, then the circular buffer is full; otherwise it is empty. In either case the
+underlying backing data is an uninitialized vector of length `capacity`.
```
Ah, it was convenient for benchmarking, but you're right that it's not useful. It's gone now.

vyu commented on exercises/circular-buffer/example.jl:
```diff
@@ -0,0 +1,177 @@
+    CircularBuffer(v::Vector{T}) where {T} = (n = length(v); new{T}(n, n, v))
+end
+
+CircularBuffer{T}(v::Vector{T}) where {T} = CircularBuffer(v::Vector{T})
```
This is no longer important now that I've removed it, but I don't think having both constructors is redundant? Unless you mean the type parameter itself. I agree that explicitly calling the constructor with `T` is redundant, but even so I still needed to define both constructors in order for `CircularBuffer(v)` and `CIrcularBuffer{T}(v)` (where `v` is a `Vector{T}`) to work identically. I don't know a way to do that with a single constructor.

vyu commented:
> I'm not convinced that we need all these examples and they make the PR harder to review.

You're right, I was having trouble looking through the PR too. :smile: I've removed the directory of examples (they can still be generated with `generate_defaults.jl`).

> How about the tool generates the tests for only the exercises named in a config file and puts those directly into the exercises/ dir (without symlinking)?
>
> I don't really think we need regression testing here, but if you'd particularly like them then I think we'll be fine with regression testing just the one exercise, not all possible exercises.

I've renamed the script `bin/diff-generated-runtests.sh` to `bin/generate-runtests.sh`. It now generates `runtests.jl` for the slugs listed in the file (currently just circular-buffer and darts) and writes them into `exercises/$slug`. I think it's fine to have the list in the script rather than a separate config file?

I've committed all the changes I have in mind now.

cmcaine commented:
> I've renamed the script bin/diff-generated-runtests.sh to bin/generate-runtests.sh. It now generates runtests.jl for the slugs listed in the file (currently just circular-buffer and darts) and writes them into exercises/$slug. I think it's fine to have the list in the script rather than a separate config file?

Sounds fine to me.

> Thanks for the review! Regarding circular-buffer, what do you think about removing all the bonus tasks and tests? If it's not a good extension to the exercise, I don't mind removing it.

I think something like this would be sufficient:

>> This exercise is quite large and potentially complicated, and that makes it more challenging and time-consuming to mentor. To help your mentor out, please don't submit code for the bonus exercises until your mentor has reviewed your solution for the first part of the exercise.

cmcaine commented on exercises/circular-buffer/.meta/hints.md:
```diff
@@ -0,0 +1,70 @@
+[`Base`](https://github.com/JuliaLang/julia/tree/master/base) module to see function
+definitions and figure out which ones to extend. To locate the relevant code for a function
+call, you can use the
+[@which](https://docs.julialang.org/en/v1/stdlib/InteractiveUtils/#InteractiveUtils.@which)
```
```suggestion
[`@which`](https://docs.julialang.org/en/v1/stdlib/InteractiveUtils/#InteractiveUtils.@which)
```

SaschaMann commented:
I wonder if the GeneratorUtils package should be moved to a separate repo so that it can be a "proper" package. What do you think? We could create exercism/GeneratorUtils.jl for that

cmcaine commented:
I don't think we gain anything from that.

On Sun, 28 Jun 2020, 15:52 Sascha Mann, <notifications@github.com> wrote:

> I wonder if the GeneratorUtils package should be moved to a separate repo
> so that it can be a "proper" package. What do you think? We could create
> exercism/GeneratorUtils.jl for that
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/exercism/julia/pull/251#issuecomment-650773929>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABNZA6IVZOV4U6CHSRT3LLTRY5KMRANCNFSM4OJBG6LQ>
> .
>

SaschaMann commented:
Easier to get rendered docs, but I'm fine with leaving it here, too.

(I've edited your comments to remove the email quote, it made browsing this PR through the app really annoying because it's not hidden in a spoiler there)

vyu commented:
> I think something like this would be sufficient:
>
> > This exercise is quite large and potentially complicated, and that makes it more challenging and time-consuming to mentor. To help your mentor out, please don't submit code for the bonus exercises until your mentor has reviewed your solution for the first part of the exercise.

Sounds good, added.

I've also added a simpler example to circular-buffer for the exercise without bonuses. I hope it's useful to have.

> I wonder if the GeneratorUtils package should be moved to a separate repo so that it can be a "proper" package. What do you think? We could create exercism/GeneratorUtils.jl for that

I don't know, I'll leave it to you all to decide!

cmcaine commented:
Oh, the practice in this repo is to rebase, squash and then merge with fast-forwarding, for whatever reason.

In that context, it would be nice if this is rebased and squashed into three commits along the obvious lines at some point after Sascha has a look.

SaschaMann commented:
> Any objections?

None

SaschaMann commented:
Thanks! this looks quite useful :)

vyu commented:
Thanks both for your help! @cmcaine I really appreciate the time you took to review and comment.
